test/e2e/test/apmserver/checks_apm.go

```diff
@@ -181,47 +224,55 @@ type CountResult struct {
 
 // CheckEventsInElasticsearch checks that the events sent in the previous step have been stored.
 // We only count document to not rely on the internal schema of the APM Server.
-func (c *apmClusterChecks) CheckEventsInElasticsearch(apm apmv1.ApmServer, k *test.K8sClient) test.Step {
-	return test.Step{
-		Name: "Events should eventually show up in Elasticsearch",
-		Test: test.Eventually(func() error {
-			// Fetch the last version of the APM Server
-			var updatedApmServer apmv1.ApmServer
-			if err := k.Client.Get(context.Background(), k8s.ExtractNamespacedName(&apm), &updatedApmServer); err != nil {
-				return err
-			}
+func (c *apmClusterChecks) checkEventsInElasticsearch(apm apmv1.ApmServer, k *test.K8sClient) error {
+	var updatedApmServer apmv1.ApmServer
+	if err := k.Client.Get(context.Background(), k8s.ExtractNamespacedName(&apm), &updatedApmServer); err != nil {
+		return err
+	}
 
-			if !updatedApmServer.Spec.ElasticsearchRef.IsDefined() {
-				// No ES is referenced, do not try to check data
-				return nil
-			}
+	if !updatedApmServer.Spec.ElasticsearchRef.IsDefined() {
+		// No ES is referenced, do not try to check data
+		return nil
+	}
 
-			v, err := version.Parse(updatedApmServer.Spec.Version)
-			if err != nil {
-				return err
-			}
+	metricIndex, errorIndex, err := getIndexNames(updatedApmServer)
+	if err != nil {
+		return err
+	}
 
-			// Check that the metric and error have been stored
-			// default to indices names from 6.x
-			metricIndex := fmt.Sprintf("apm-%s-2017.05.30", updatedApmServer.EffectiveVersion())
-			errorIndex := fmt.Sprintf("apm-%s-2018.08.09", updatedApmServer.EffectiveVersion())
-			switch v.Major {
-			case 7:
-				metricIndex = fmt.Sprintf("apm-%s-metric-2017.05.30", updatedApmServer.EffectiveVersion())
-				errorIndex = fmt.Sprintf("apm-%s-error-2018.08.09", updatedApmServer.EffectiveVersion())
-			case 8:
-				// these are datastreams and not indices, but can be searched/counted in the same way
-				metricIndex = "metrics-apm.app.1234_service_12a3-default"
-				errorIndex = "logs-apm.error-default"
-			}
+	return retry.UntilSuccess(func() error {
```

Comment:
I'm not sure to understand why we are retrying during 5 minutes? What external event would unlock this situation? My understanding is that if `assertCountIndexEqual` fails, it is likely because `checkEventsAPI` (`c.apmClient.IntakeV2Events`) has also failed?

Reply:
Hopefully this update will show what's going on in some detail.

If I move the retry around this block and add some logging, we can see what failures we're working with

```
			return retry.UntilSuccess(func() error {
				if err := c.checkEventsAPI(apm); err != nil {
					fmt.Printf("checkEventsApi failure: %s\n", err)
					return err
				}
				if err := c.checkEventsInElasticsearch(apm, k); err != nil {
					fmt.Printf("checkEventsInElasticsearch failure: %s\n", err)
					return err
				}
				return nil
			}, 5*time.Minute, 10*time.Second)
```

With this change, here's the result.  It fails before even attempting the version upgrade.

1st attempt:  The messages are sent to apmserver successfully, but they haven't made it all the way to ES yet, as the metric index hasn't even been created yet.
2nd attempt: Before we attempt to send another message, we count the messages in the index (metric) and find 1 (which likely is the first message we attempted to send.  Messages sent successfully to apmserver again, but we are expecting 2 messages in the index, but only find one (again).
3rd attempt: This process keeps repeating.

Seems like we aren't waiting long enough for the message to get from apm -> ES.

```
EventsInElasticsearch failure: elasticsearch client failed for https://apmserver-upgrade-rxzf-es-internal-http.e2e-mercury.svc:9200/apm-7.17.1-metric-2017.05.30/_count: 404 Not Found: {Status:404 Error:{CausedBy:{Reason: Type:} Reason:no such index [apm-7.17.1-metric-2017.05.30] Type:index_not_found_exception RootCause:[{Reason:no such index [apm-7.17.1-metric-2017.05.30] Type:index_not_found_exception}]}}\n"}
{"Time":"2022-04-28T10:44:56.167184-05:00","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/apm","Test":"TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch","Output":"checkEventsInElasticsearch failure: 2 documents expected in index apm-7.17.1-metric-2017.05.30, got 1 instead\n"}
{"Time":"2022-04-28T10:45:06.47493-05:00","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/apm","Test":"TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch","Output":"checkEventsInElasticsearch failure: 3 documents expected in index apm-7.17.1-metric-2017.05.30, got 2 instead\n"}
{"Time":"2022-04-28T10:45:16.885698-05:00","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/apm","Test":"TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch","Output":"checkEventsInElasticsearch failure: 4 documents expected in index apm-7.17.1-metric-2017.05.30, got 3 instead\n"}
{"Time":"2022-04-28T10:45:27.190003-05:00","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/apm","Test":"TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch","Output":"checkEventsInElasticsearch failure: 5 documents expected in index apm-7.17.1-metric-2017.05.30, got 4 instead\n"}
{"Time":"2022-04-28T10:45:37.492295-05:00","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/apm","Test":"TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch","Output":"checkEventsInElasticsearch failure: 6 documents expected in index apm-7.17.1-metric-2017.05.30, got 5 instead\n"}
{"Time":"2022-04-28T10:45:47.799507-05:00","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/apm","Test":"TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch","Output":"checkEventsInElasticsearch failure: 7 documents expected in index apm-7.17.1-metric-2017.05.30, got 6 instead\n"}
{"Time":"2022-04-28T10:45:58.102207-05:00","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/apm","Test":"TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch","Output":"checkEventsInElasticsearch failure: 8 documents expected in index apm-7.17.1-metric-2017.05.30, got 7 instead\n"}
{"Time":"2022-04-28T10:46:08.396236-05:00","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/apm","Test":"TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch","Output":"checkEventsInElasticsearch failure: 9 documents expected in index apm-7.17.1-metric-2017.05.30, got 8 instead\n"}
```

If I change things, and retry the counting of the indexes after dropping messages in apmserver, things succeed again.

```
			if err := c.checkEventsAPI(apm); err != nil {
				fmt.Printf("checkEventsApi failure: %s\n", err)
				return err
			}
			return retry.UntilSuccess(func() error {
				if err := c.checkEventsInElasticsearch(apm, k); err != nil {
					fmt.Printf("checkEventsInElasticsearch failure: %s\n", err)
					return err
				}
				return nil
			}, 5*time.Minute, 10*time.Second)
```

Errors during this run

Before version upgrade we are missing the index, but the retry after 10 seconds succeeds.
```
{"Time":"2022-04-28T11:56:34.904163-05:00","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/apm","Test":"TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch","Output":"Retries (30m0s timeout): .checkEventsInElasticsearch failure: elasticsearch client failed for https://apmserver-upgrade-hdgx-es-internal-http.e2e-mercury.svc:9200/apm-7.17.1-metric-2017.05.30/_count: 404 Not Found: {Status:404 Error:{CausedBy:{Reason: Type:} Reason:no such index [apm-7.17.1-metric-2017.05.30] Type:index_not_found_exception RootCause:[{Reason:no such index [apm-7.17.1-metric-2017.05.30] Type:index_not_found_exception}]}}\n"}
```

After the version upgrade, we get some failures, but eventually succeed

1: index initially not found
```
{"Time":"2022-04-28T12:05:02.499518-05:00","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/apm","Test":"TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch#01","Output":"Retries (30m0s timeout): .checkEventsInElasticsearch failure: elasticsearch client failed for https://apmserver-upgrade-hdgx-es-internal-http.e2e-mercury.svc:9200/metrics-apm.app.1234_service_12a3-default/_count: 404 Not Found: {Status:404 Error:{CausedBy:{Reason: Type:} Reason:no such index [metrics-apm.app.1234_service_12a3-default] Type:index_not_found_exception RootCause:[{Reason:no such index [metrics-apm.app.1234_service_12a3-default] Type:index_not_found_exception}]}}\n"}
```

This repeats from `2022-04-28T 12:05:02.499518-05:00` until `2022-04-28T 12:10:06.010821-05:00` (about 5 mins, which is the time for ES to re-read it's file-based users/roles IIRC) when it finally succeeds.

Attempt 3

Sleep for 10 seconds in between inserting message(s), and checking to see if they made it into ES.

```
			return retry.UntilSuccess(func() error {
				if err := c.checkEventsAPI(apm); err != nil {
					fmt.Printf("checkEventsApi failure: %s\n", err)
					return err
				}
				time.Sleep(10*time.Second)
				if err := c.checkEventsInElasticsearch(apm, k); err != nil {
					fmt.Printf("checkEventsInElasticsearch failure: %s\n", err)
					return err
				}
				return nil
			}, 5*time.Minute, 10*time.Second)
```

Before version upgrade, the first attempt to insert + read messages through apmserver succeeds without failure.

After version upgrade, we get 2x index not found failures, and then it succeeds within < 1 minute (did we just get lucky with timing that it didn't take 5 mins?)

```
{"Time":"2022-04-28T13:37:28.722225-05:00","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/apm","Test":"TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch#01","Output":"Retries (30m0s timeout): .checkEventsInElasticsearch failure: elasticsearch client failed for https://apmserver-upgrade-nkvk-es-internal-http.e2e-mercury.svc:9200/metrics-apm.app.1234_service_12a3-default/_count: 404 Not Found: {Status:404 Error:{CausedBy:{Reason: Type:} Reason:no such index [metrics-apm.app.1234_service_12a3-default] Type:index_not_found_exception RootCause:[{Reason:no such index [metrics-apm.app.1234_service_12a3-default] Type:index_not_found_exception}]}}\n"}
{"Time":"2022-04-28T13:37:49.022055-05:00","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/apm","Test":"TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch#01","Output":"checkEventsInElasticsearch failure: elasticsearch client failed for https://apmserver-upgrade-nkvk-es-internal-http.e2e-mercury.svc:9200/metrics-apm.app.1234_service_12a3-default/_count: 404 Not Found: {Status:404 Error:{CausedBy:{Reason: Type:} Reason:no such index [metrics-apm.app.1234_service_12a3-default] Type:index_not_found_exception RootCause:[{Reason:no such index [metrics-apm.app.1234_service_12a3-default] Type:index_not_found_exception}]}}\n"}
{"Time":"2022-04-28T13:38:09.523725-05:00","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/apm","Test":"TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch#01","Output":"\n"}
```

So from what I see, we have 2 options:

1. Retry the attempt to find the messages in ES for a bit until we fail/succeed.
2. Sleep a bit between initial count+insert, and the subsequent count to see if the message(s) arrived successfully.

@pebrc @barkbay Please let me know what you think, and which direction you'd prefer me take this, and I'll get it implemented, and we can finish off this PR.

Thanks!

Reply:
I also added some debug logs to make it clear what is happening in my case and with your original proposal (516df86b2):

```
=== RUN   TestAPMServerVersionUpgradeToLatest8x/ApmServer_should_accept_event_and_write_data_to_Elasticsearch#01
Retries (5m0s timeout):
04-29-2022 13:43:54 - [DEBUG] Entering checkEventsAPI(...)
04-29-2022 13:43:56 - [DEBUG] State before injecting events: c.metricIndexCount=0 c.errorIndexCount=0
04-29-2022 13:43:56 - [DEBUG] Injecting event using c.apmClient.IntakeV2Events
04-29-2022 13:43:56 - [DEBUG] Leaving checkEventsAPI without error, c.apmClient.IntakeV2Events seems to be successful
04-29-2022 13:43:56 - [DEBUG] Entering checkEventsInElasticsearch(...)
04-29-2022 13:43:56 - [DEBUG] entering assertCountIndexEqual for index metrics-apm.app.1234_service_12a3-default, expecting 1 events
04-29-2022 13:43:56 - [ERROR] leaving assertCountIndexEqual with error: 404 Not Found: no such index [metrics-apm.app.1234_service_12a3-default]
.... 5 minutes later, still the same error ...
04-29-2022 13:48:48 - [DEBUG] entering assertCountIndexEqual for index metrics-apm.app.1234_service_12a3-default, expecting 1 events
04-29-2022 13:48:48 - [ERROR] leaving assertCountIndexEqual with error: 404 Not Found: no such index [metrics-apm.app.1234_service_12a3-default]

    step.go:43: 
        	Error Trace:	utils.go:88
        	Error:      	Received unexpected error:
        	            	timeout reached after 5m0s
```

The index is never created, `checkEventsAPI` is never retried.
I revisited the original retry policy (based on 516df86b2) to give a chance to `checkEventsAPI(...)` to be retried:

```diff
--- a/test/e2e/test/apmserver/checks_apm.go
+++ b/test/e2e/test/apmserver/checks_apm.go
@@ -246,7 +246,9 @@ func (c *apmClusterChecks) checkEventsInElasticsearch(apm apmv1.ApmServer, k *te
                }
 
                return assertCountIndexEqual(c.esClient, errorIndex, c.errorIndexCount+1)
-       }, 5*time.Minute, 10*time.Second)
+               // We need some retries here as the events might not have been indexed in Elasticsearch.
+               // timeout must be less than the parent retry policy, but long enough to let events flow to Elasticsearch.
+       }, 30*time.Second, 2*time.Second)
 }
```

With that change applied it works after 3 calls to `checkEventsAPI()`:

```
# First attempt: KO with `404`, even after the updated timeout of `30*time.Second` index is not created
Retries (5m0s timeout): .04-29-2022 12:57:21 - [DEBUG] Entering checkEventsAPI(...)
04-29-2022 12:57:23 - [DEBUG] State before injecting events: c.metricIndexCount=0 c.errorIndexCount=0
04-29-2022 12:57:23 - [DEBUG] Injecting event using c.apmClient.IntakeV2Events
04-29-2022 12:57:23 - [DEBUG] Leaving checkEventsAPI without error, c.apmClient.IntakeV2Events seems to be successful
04-29-2022 12:57:23 - [DEBUG] Entering checkEventsInElasticsearch(...)
04-29-2022 12:57:23 - [DEBUG] entering assertCountIndexEqual for index metrics-apm.app.1234_service_12a3-default, expecting 1 events
04-29-2022 12:57:23 - [ERROR] leaving assertCountIndexEqual with error: 404 Not Found:no such index [metrics-apm.app.1234_service_12a3-default]
04-29-2022 12:57:25 - [DEBUG] entering assertCountIndexEqual for index metrics-apm.app.1234_service_12a3-default, expecting 1 events
[ ... last error x 12 ... ]
04-29-2022 12:57:52 - [DEBUG] entering assertCountIndexEqual for index metrics-apm.app.1234_service_12a3-default, expecting 1 events
04-29-2022 12:57:52 - [ERROR] leaving assertCountIndexEqual with error: 404 Not Found:no such index [metrics-apm.app.1234_service_12a3-default]

# Second attempt: same problem, KO with `404`
.04-29-2022 12:57:56 - [DEBUG] Entering checkEventsAPI(...)
04-29-2022 12:57:56 - [DEBUG] State before injecting events: c.metricIndexCount=0 c.errorIndexCount=0
04-29-2022 12:57:56 - [DEBUG] Injecting event using c.apmClient.IntakeV2Events
04-29-2022 12:57:56 - [DEBUG] Leaving checkEventsAPI without error, c.apmClient.IntakeV2Events seems to be successful
04-29-2022 12:57:56 - [DEBUG] Entering checkEventsInElasticsearch(...)
04-29-2022 12:57:56 - [DEBUG] entering assertCountIndexEqual for index metrics-apm.app.1234_service_12a3-default, expecting 1 events
04-29-2022 12:57:56 - [ERROR] leaving assertCountIndexEqual with error: 404 Not Found:no such index [metrics-apm.app.1234_service_12a3-default]
04-29-2022 12:57:58 - [DEBUG] entering assertCountIndexEqual for index metrics-apm.app.1234_service_12a3-default, expecting 1 events
[ last error x 12]
04-29-2022 12:58:25 - [DEBUG] entering assertCountIndexEqual for index metrics-apm.app.1234_service_12a3-default, expecting 1 events
04-29-2022 12:58:25 - [ERROR] leaving assertCountIndexEqual with error: 404 Not Found:no such index [metrics-apm.app.1234_service_12a3-default]

# Third attempt: OK after 1 `404`
04-29-2022 12:58:29 - [DEBUG] Entering checkEventsAPI(...)
04-29-2022 12:58:29 - [DEBUG] State before injecting events: c.metricIndexCount=0 c.errorIndexCount=0
04-29-2022 12:58:29 - [DEBUG] Injecting event using c.apmClient.IntakeV2Events
04-29-2022 12:58:30 - [DEBUG] Leaving checkEventsAPI without error, c.apmClient.IntakeV2Events seems to be successful
04-29-2022 12:58:30 - [DEBUG] Entering checkEventsInElasticsearch(...)
04-29-2022 12:58:30 - [DEBUG] entering assertCountIndexEqual for index metrics-apm.app.1234_service_12a3-default, expecting 1 events
04-29-2022 12:58:30 - [ERROR] leaving assertCountIndexEqual with error: 404 Not Found:no such index [metrics-apm.app.1234_service_12a3-default]
04-29-2022 12:58:32 - [DEBUG] entering assertCountIndexEqual for index metrics-apm.app.1234_service_12a3-default, expecting 1 events
04-29-2022 12:58:32 - [DEBUG] 1 documents expected in index metrics-apm.app.1234_service_12a3-default, got 0 instead
04-29-2022 12:58:34 - [DEBUG] entering assertCountIndexEqual for index metrics-apm.app.1234_service_12a3-default, expecting 1 events
04-29-2022 12:58:34 - [DEBUG] leaving assertCountIndexEqual without error
04-29-2022 12:58:34 - [DEBUG] entering assertCountIndexEqual for index logs-apm.error-default, expecting 1 events
04-29-2022 12:58:34 - [DEBUG] leaving assertCountIndexEqual without error
```

Reply:
> So from what I see, we have 2 options:

A third option would be to say that despite our efforts it is very hard to state that an APM upgrade is successful, we could also use an annotation in the `podTemplate` to trigger a restart and validate that `observedGeneration` is updated as expected.

Reply:
> > about 5 mins, which is the time for ES to re-read it's file-based users/roles IIRC
> 
> I would be really surprised if ES only reloads these files every 5 minutes (that would not match my experience).

@barkbay https://www.elastic.co/guide/en/elasticsearch/reference/current/file-realm.html#file-realm
bottom of page, every 5 `seconds`.  Well, I was quite the order of magnitude off... 

I'm going to update this with these changes

```
}, 30*time.Second, 2*time.Second)
```

as you're suggesting, and run some additional tests.